src/com/palmergames/bukkit/towny/object/Resident.java

```diff
@@ -307,24 +310,69 @@ public Town getTown() throws NotRegisteredException {
 
 	public void setTown(Town town) throws AlreadyRegisteredException {
 
+		if (this.town == town)
+			return;
+
+		Towny.getPlugin().deleteCache(this.getName());
+		setTitle("");
+		setSurname("");
+		updatePerms();
+		
+
 		if (town == null) {
 			this.town = null;
-			setTitle("");
-			setSurname("");
-			updatePerms();
 			return;
 		}
 
-		if (this.town == town)
-			return;
-
 		if (hasTown())
 			throw new AlreadyRegisteredException();
 
 		this.town = town;
-		setTitle("");
-		setSurname("");
-		updatePerms();
+		town.addResident(this);
+		BukkitTools.getPluginManager().callEvent(new TownAddResidentEvent(this, town));
+	}
+	
+	public void removeTown() {
+		
+		if (!hasTown())
+			return;
+
+		Town town = this.town;
+		
+		BukkitTools.getPluginManager().callEvent(new TownRemoveResidentEvent(this, town));
+		try {
+			
+			town.removeResident(this);
+			
+		} catch (NotRegisteredException e1) {
+			e1.printStackTrace();
+		} catch (EmptyTownException e1) {
+			TownyUniverse.getInstance().getDataSource().removeTown(town);
+		}
+		
+		for (TownBlock townBlock : new ArrayList<>(townBlocks)) {
```

Comment:
This should be just `townsblocks` not `new ArrayList<>(townBlocks)`

Reply:
I started with that and it throws concurrentmodification exceptions all over.

Reply:
Fixed via iterator